src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs

```diff
@@ -93,5 +115,80 @@ public async Task<IModalProgressDialogSession> StartModalProgressDialogAsync(str
             var session = waitForDialogFactory.StartWaitDialog(caption, progressData);
             return new VisualStudioProgressDialogSession(session);
         }
+
+        public static async ValueTask<NuGetUIContext> CreateAsync(
+            ISourceRepositoryProvider sourceRepositoryProvider,
+            ISettings settings,
+            IVsSolutionManager solutionManager,
+            IPackageRestoreManager packageRestoreManager,
+            IOptionsPageActivator optionsPageActivator,
+            IUserSettingsManager userSettingsManager,
+            IDeleteOnRestartManager deleteOnRestartManager,
+            IEnumerable<IVsPackageManagerProvider> packageManagerProviders,
+            INuGetLockService lockService,
+            CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(sourceRepositoryProvider);
+            Assumes.NotNull(settings);
+            Assumes.NotNull(solutionManager);
+            Assumes.NotNull(packageRestoreManager);
+            Assumes.NotNull(optionsPageActivator);
+            Assumes.NotNull(userSettingsManager);
+            Assumes.NotNull(deleteOnRestartManager);
+            Assumes.NotNull(packageManagerProviders);
+            Assumes.NotNull(lockService);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            IServiceBroker serviceBroker = await BrokeredServicesUtilities.GetRemoteServiceBrokerAsync();
+
+            var solutionManagerService = new NuGetSolutionManagerServiceWrapper()
+            {
+                Service = await GetSolutionManagerServiceAsync(cancellationToken)
+            };
+
+            var packageManager = new NuGetPackageManager(
+                sourceRepositoryProvider,
+                settings,
+                solutionManager,
+                deleteOnRestartManager);
+
+            var actionEngine = new UIActionEngine(
+                sourceRepositoryProvider,
+                packageManager,
+                lockService);
+
+            return new NuGetUIContext(
+                sourceRepositoryProvider,
+                serviceBroker,
+                solutionManager,
+                solutionManagerService,
+                packageManager,
+                actionEngine,
+                packageRestoreManager,
+                optionsPageActivator,
+                userSettingsManager,
+                packageManagerProviders);
+        }
+
+        private void OnAvailabilityChanged(object sender, BrokeredServicesChangedEventArgs e)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    _solutionManagerService.Service = await GetSolutionManagerServiceAsync(CancellationToken.None);
```

Comment:
>   _solutionManagerService.Service [](start = 17, length = 34)

Before we replace this, should we Dispose of it first?

Reply:
Yep!